src/core/pointcloud/qgspointcloudstatscalculator.cpp

```diff
@@ -0,0 +1,268 @@
+#include <QtConcurrent/QtConcurrentMap>
+
+struct StatsProcessor
+{
+    typedef QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> result_type;
+
+    StatsProcessor( QgsPointCloudIndex *index, QgsPointCloudRequest request, QgsFeedback *feedback )
+      : mIndex( index ), mRequest( request ), mFeedback( feedback )
+    {
+
+    }
+
+    QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> operator()( IndexedPointCloudNode node )
+    {
+      QMap<QString, QgsPointCloudStatsCalculator::AttributeStatistics> statsMap;
+      for ( QgsPointCloudAttribute attribute : mRequest.attributes().attributes() )
+      {
+        QgsPointCloudStatsCalculator::AttributeStatistics summary;
+        summary.minimum = std::numeric_limits<double>::max();
+        summary.maximum = std::numeric_limits<double>::lowest();
+        summary.count = 0;
+        summary.classCount.clear();
+        statsMap[ attribute.name() ] = summary;
+      }
+
+      std::unique_ptr<QgsPointCloudBlock> block;
+      if ( mIndex->accessType() == QgsPointCloudIndex::Local )
+      {
+        block.reset( mIndex->nodeData( node, mRequest ) );
+      }
+      else
+      {
+        QgsPointCloudBlockRequest *request = mIndex->asyncNodeData( node, mRequest );
+        QEventLoop loop;
+        QObject::connect( request, &QgsPointCloudBlockRequest::finished, &loop, &QEventLoop::quit );
+        QObject::connect( mFeedback, &QgsFeedback::canceled, &loop, &QEventLoop::quit );
+        loop.exec();
+        if ( !mFeedback->isCanceled() )
+          block.reset( request->block() );
+        if ( !request->block() )
+        {
+          QgsMessageLog::logMessage( QObject::tr( "Unable to calculate statistics for node %1, error: \"%2\"" ).arg( node.toString() ).arg( request->errorStr() ) );
+          return result_type();
+        }
+      }
+
+      if ( !block.get() )
+      {
+        return result_type();
+      }
+
+      const char *ptr = block->data();
+      int count = block->pointCount();
+      const QgsPointCloudAttributeCollection attributes = block->attributes();
+      int recordSize = attributes.pointRecordSize();
+
+      for ( int i = 0; i < count; ++i )
+      {
+        for ( QgsPointCloudAttribute attribute : attributes.attributes() )
+        {
+          if ( mFeedback->isCanceled() )
+          {
+            return result_type();
+          }
+          double attributeValue = 0;
+          int attributeOffset = 0;
+          attributes.find( attribute.name(), attributeOffset );
+
+          QgsPointCloudStatsCalculator::AttributeStatistics &stats = statsMap[ attribute.name() ];
+          QgsPointCloudRenderContext::getAttribute( ptr, i * recordSize + attributeOffset, attribute.type(), attributeValue );
+          stats.minimum = std::min( stats.minimum, attributeValue );
+          stats.maximum = std::max( stats.maximum, attributeValue );
+          stats.count++;
+          if ( attribute.name() == QStringLiteral( "Classification" ) )
```

Comment:
it would be more efficient to just find out beforehand which attribute index is the one with classification, to avoid lots of string comparisons...